### DISCUSSION TITLE
How to send a request directly to the language server set up by coc.nvim

### DISCUSSION THREAD

**Discussion by uga-rosa** [neoclide/coc.nvim#4662] [Q&A] [answered]
I would like to send a request like `textDocument/references` directly from the vim script and receive the resulting `Location[]`.
If you know neovim's built-in lsp client, interpret this as wanting to do something like `vim.lsp.buf_request_sync()`.

**Comment by fannheyward** [comment]
`:h CocAction('references')`.

**Reply by uga-rosa** [reply]
I would like to touch a lower layer. The current cursor position is useless since it is for use in the fuzzy finder. I would like to use all the parameters that I have created to send to LS. Is there a way?
Would it make sense if I say I want to access `textDocument/references`, not `references`?

**Reply by fannheyward** [reply, marked as answer]
> want to access `textDocument/references`, not `references`

They were same, `references` did `ReferencesRequest`.

Use `:h CocRequest()`, with langserver id, `textDocument/references`, and custom position params.

**Reply by fannheyward** [reply]
Also `:h CocRequestAsync()` for async requests.

**Reply by uga-rosa** [reply]
Thanks!

**Reply by uga-rosa** [reply]
Can I get a list of the langserver ids that are enabled for a given buffer?

**Reply by fannheyward** [reply]
`:CocList services`, this lists all activated services, not current buffer only.

**Reply by uga-rosa** [reply]
Hmmm, I can't use it from the code, can I? So, does the user have to specify which LS to use?
Also, in the example above, coc-tslint is written as if its id is tslint, but where can I find this name? Does it mean that the id of coc-foo is foo or can it be found somewhere else?

**Reply by fannheyward** [reply]
> use it from the code

You mean `:CocList services` from code? Try `:call CocAction('services')`, with id that you want.

**Reply by fannheyward** [reply]
The `id` is registered by extensions or coc-settings.json, coc-foo is foo, but not must.

**Reply by uga-rosa** [reply]
Thanks, I was trying with coc-deno, but it seems to be another problem that it doesn't appear in services, I tried with coc-lua and I could see the id properly.

**Reply by fannheyward** [reply]
coc-deno's issue, coc-deno starts the client, but not register to services, will fix this later.

**Reply by fannheyward** [reply]
Fixed with coc-deno v3.12.0
